File: BOJ/완전탐색/14500/14500.py
# 14500 테트로미노
# https://www.acmicpc.net/problem/14500
import sys
input = sys.stdin.readline

res = 0


def solution(N, M, board):
    answer = 0
    visited = [[False] * M for _ in range(N)]  # 방문리스트
    dir = [[1, 0], [-1, 0], [0, -1], [0, 1]]

    def specialShape(x, y):
        global res

        for i in range(4):
            tot = board[x][y]
            for j in range(3):
                k = (i+j) % 4
                nx = x + dir[k][0]
                ny = y + dir[k][1]

                if nx < 0 or ny < 0 or nx >= N or ny >= M:
                    continue

                tot += board[nx][ny]
            res = max(res, tot)
        return

    def commonShape(x, y, tot, depth):
        global res
        if depth == 3:
            res = max(res, tot)
            return

        for d in dir:
            nx = x + d[0]
            ny = y + d[1]

            if nx < 0 or ny < 0 or nx >= N or ny >= M:
                continue

            if visited[nx][ny]:
                continue

            visited[nx][ny] = True
            commonShape(nx, ny, tot+board[nx][ny], depth+1)
            visited[nx][ny] = False
        return

    # 완전탐색 시작
    for i in range(N):
        for j in range(M):
            specialShape(i, j)
            visited[i][j] = True
            commonShape(i, j, board[i][j], 0)
            visited[i][j] = False

    answer = res
    return answer


N, M = map(int, input().split())
board = [list(map(int, input().split())) for _ in range(N)]
print(solution(N, M, board))

File: BOJ/완전탐색/14500/14500.cpp
// 14500 테트로미노
// https://www.acmicpc.net/problem/14500
#include <iostream>

using namespace std;

int N, M, res = -1;
int board[501][501];
int visited[501][501];
int dir[4][2] = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};

int mx(int a, int b) {
    return a > b ? a : b;
}

void isSpecialShape(int x, int y) {
    // ㅜ
    if (x+1 < N && y + 2 < M)
    {
        res = mx(res, board[x][y]+board[x][y+1]+board[x+1][y+1]+board[x][y+2]);
    }

    // ㅏ
    if (x+2 < N && y+1 < M)
    {
        res = mx(res, board[x][y]+board[x+1][y]+board[x+1][y+1]+board[x+2][y]);
    }

    // ㅓ
    if (x+2 < N && y - 1 >= 0)
    {
        res = mx(res, board[x][y]+board[x+1][y]+board[x+1][y-1]+board[x+2][y]);
    }

    // ㅗ
    if (x-1 >= 0 && y +2 < M)
    {
        res = mx(res, board[x][y]+board[x][y+1]+board[x-1][y+1]+board[x][y+2]);
    }
    return;
}

void DFS(int x, int y, int cellTotal, int depth) {
    if (depth == 4)
    {
        res = mx(cellTotal, res);
        return;
    }

    for (auto& d: dir)
    {
        int nx = x + d[0];
        int ny = y + d[1];

        if (nx < 0 || ny < 0 || nx >= N || ny >= M) continue;

        if (visited[nx][ny]) continue;

        visited[nx][ny] = 1;
        DFS(nx, ny, cellTotal+board[nx][ny], depth+1);
        visited[nx][ny] = 0;
    }
    return;
}

int solution(void) {
    int answer = 0;

    for (int i = 0; i < N; ++i)
    {
        for (int j = 0; j < M; ++j)
        {
            visited[i][j] = 1;
            isSpecialShape(i,j);
            DFS(i,j,board[i][j], 1);
            visited[i][j] = 0;
        }
    }

    answer = res;
    return answer;
}

int main(void) {
    ios::sync_with_stdio(0);
    cin.tie(0);
    cout.tie(0);

    cin >> N >> M;
    for (int i = 0; i < N; ++i)
    {
        for (int j = 0; j < M; ++j)
        {
            cin >> board[i][j];
        }
    }

    cout << solution();
    return 0;
}